### PR TITLE
Cache full-view terminal content

### LINKED ISSUE #284: [Feature]: Cache full-screen terminal content to avoid unnecessary full-view rerasterization
**Title:** [Feature]: Cache full-screen terminal content to avoid unnecessary full-view rerasterization

**Labels:** feature, medium

## Status Quo

Grid mode already uses a per-session texture cache in `src/render/renderer.zig`: a session is rerendered into its cache texture only when its `render_epoch` changes. Full-screen mode does not use the same optimization path. When a full-view frame is rendered, Architect walks the visible terminal cells and draws content again even if terminal content is unchanged and only overlays or event processing triggered the frame.

## Objectives

Make full-screen terminal interaction more efficient by reusing previously rendered terminal content whenever possible. The user should get the same visuals and behavior as today, but Architect should spend less time rescanning and rerendering the full terminal when nothing in the terminal content changed.

## User Flow

**Trigger:** The user works in a full-screen terminal view.

1. Architect renders the full-screen terminal content into a reusable cache.
2. If terminal content has not changed, later frames reuse that cached texture.
3. If terminal content changes, the cache is refreshed and the updated content is presented.

**Result:** Full-screen terminal interaction feels lighter and more consistent, especially when UI or event activity causes extra frames.

## Scope

**In scope:**

- Add a cached full-view rendering path that follows the existing session dirty model.
- Preserve overlays, cursor rendering, selection rendering, and visual parity with the current full-view renderer.

**Out of scope:**

- Fine-grained row or rectangle damage tracking inside the terminal surface.
- Reworking grid-mode caching beyond what is needed for shared abstractions.

### Affected Modules

- `src/render/renderer.zig`: introduce a cached full-view path and unify cache handling where it makes sense.
- `src/ui/session_view_state.zig`: adjust any view-state assumptions if cache invalidation needs explicit overlay separation.
- `src/metrics.zig`: optionally expose counters useful for cache-hit verification.

### Tasks

1. [ ] Refactor renderer cache handling so full-view mode can reuse the same dirty and clean model as grid mode. `src/render/renderer.zig`
2. [ ] Add a full-screen content cache path that redraws only when the focused session's content epoch changes. `src/render/renderer.zig`
3. [ ] Keep overlays and transient effects rendering on top of cached content without visual regressions. `src/render/renderer.zig`
4. [ ] Write tests for cache invalidation on terminal content changes, correct reuse on unchanged frames, and parity for cursor, selection, and scrollback cases.
5. [ ] Update `docs/ARCHITECTURE.md` to document full-view caching behavior.

### New Dependencies

None

## Acceptance Criteria

- [ ] Full-screen mode reuses cached terminal content on unchanged frames.
- [ ] Terminal content updates still appear immediately when `render_epoch` changes.
- [ ] Overlays and cursor or selection visuals remain correct.
- [ ] Tests cover the happy path and documented edge cases.
- [ ] `docs/ARCHITECTURE.md` is updated.
- [ ] Pre-commit hooks pass.

### PR DESCRIPTION
**Pull request by forketyfork**
Fixes #284

## Summary

This change brings steady-state full-screen rendering onto the same cached-content model Architect already uses for grid tiles. The renderer now reuses a cached full-view terminal texture when the focused session has not changed, while still keeping overlays live unless an active wave effect needs them baked into the cached texture.

## Changes

- Route steady-state full-view rendering through the shared cached-session path instead of rerasterizing the full terminal every frame.
- Track cache composition so the renderer can tell whether a cached texture contains terminal content only or terminal content plus overlays for wave effects.
- Add renderer tests that cover unchanged-frame reuse, epoch-based invalidation, and the overlay-baked cache case.
- Update the architecture guide to document that renderer caching now applies to both grid tiles and steady-state full view.

## Test plan

- [x] Open a terminal in full view and confirm cursor, selection, and scrollbar visuals still update correctly while terminal output is otherwise unchanged.
- [x] Trigger an attention or navigation wave in grid mode and confirm the animated session still shows the expected overlays during the wave.

### CODE REVIEW

File: docs/ARCHITECTURE.md [hunk 1]
```diff
@@ -99,7 +99,7 @@ These patterns are mandatory for all new code. They are derived from the archite
 
 1. **UI components use the vtable interface and communicate via UiAction queue.** Never mutate application state directly from a UI component. Push a `UiAction` to the queue; the main loop drains it after all component updates complete. (See ADR-003.)
 
-2. **Render invalidation uses epoch comparison.** When terminal content changes, increment `render_epoch` on the `SessionState`. The renderer checks `presented_epoch < render_epoch` and only re-renders on mismatch. Never force a full re-render. (See ADR-004.)
+2. **Render invalidation uses epoch comparison.** When terminal content changes, increment `render_epoch` on the `SessionState`. The renderer checks `presented_epoch < render_epoch` to know whether a session needs to be redrawn this frame, and cached session textures refresh when their stored epoch no longer matches the session epoch. This applies to both grid tiles and the steady-state full-screen terminal view. Never force a full re-render. (See ADR-004.)
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply docs/ARCHITECTURE.md]: The architecture wording now matches the implementation. It describes an epoch mismatch check rather than `presented_epoch < render_epoch`, which is the correct rule when the epoch uses wrapping increments.

File: src/render/renderer.zig [hunk 2]
```diff
@@ -1005,6 +1080,33 @@ test "getCellColor uses the live terminal palette for indexed colors" {
...
 }
 
+test "full-view cached frames reuse the existing texture when the epoch is unchanged" {
+    const entry = RenderCache.Entry{
+        .cache_epoch = 42,
+        .cache_composition = .content_only,
+    };
+
+    try std.testing.expect(!cacheNeedsRefresh(&entry, 42, cacheComposition(false)));
+}
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/render/renderer.zig]: I renamed the test so it describes the cache refresh predicate it actually exercises instead of implying an end-to-end full-view rendering test.

File: src/render/renderer.zig [hunk 3]
```diff
@@ -805,74 +908,46 @@ fn renderGridSessionCached(
...
-                const local_rect = Rect{ .x = 0, .y = 0, .w = rect.w, .h = rect.h };
-                try renderSessionContent(renderer, session, view, local_rect, scale, is_focused, font, term_cols, term_rows, current_time_ms, theme, ui_scale);
-                if (any_waving and render_overlays) {
-                    renderSessionOverlays(renderer, session, view, local_rect, is_focused, apply_effects, current_time_ms, true, theme, ui_scale);
-                }
-                cache_entry.cache_epoch = session.render_epoch;
-                _ = c.SDL_SetRenderTarget(renderer, null);
+            if (cacheNeedsRefresh(cache_entry, session.render_epoch, composition)) {
+                try refreshSessionCacheTexture(renderer, session, view, cache_entry, rect, scale, is_focused, apply_effects, font, term_cols, term_rows, current_time_ms, cache_overlays, composition, is_grid_view, theme, ui_scale);
             }
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/render/renderer.zig:948]: The steady-state `.Full` path now drops every unfocused session cache before rendering the focused session. That keeps full-view caching in place without retaining a full-window texture for every visited session. Panning still uses the two sessions involved in the animation, but once the view settles only the focused session keeps a cache.
